tools/coverage.sh

```diff
@@ -0,0 +1,47 @@
+#!/bin/bash
+# Clear coverage data, run tests, then generate a coverage report.
+#
+# Usage:
+#
+#   ./coverage.sh [KAFKA_VERSION [TOX]]
+#
+# KAFKA_VERSION - Version of Kafka to run integration tests against. See
+#                 KAFKA_ALL_VERS in the Makefile for a list of known versions.
+#                 If not provided, only unit tests will be run.
+#
+# TOX           - tox executable to use. Defaults to whatever is available on
+#                 your PATH.
+
+KAFKA_VERSION="$1"
+TOX="${2:-tox}"
+
+if [[ -n $KAFKA_VERSION ]]
+then
+    export KAFKA_VERSION
+else
+    echo "No Kafka version provided: only unit tests will be run."
+fi
+
+env_filter() {
+    if [[ -n $KAFKA_VERSION ]]
+    then
+        cat
+    else
+        egrep -v '[-]int'
+    fi
+}
+
+if ! "$TOX" -e cov_erase
+then
+    exit 1
+fi
+
+"$TOX" tox -l | env_filter | awk '{ print($0 "-coverage"); }' | xargs -n1 "$TOX" -e
```

Comment:
This is doing: use tox to list the environments, if KAFKA_VERSION is not set, strip out the lines matching '[-]int', add -coverage to it and run tox for each of those environments?
`tox -l` doesn't emit the environments which are '<foo>-coverage'?  I guess because it's not a "real" environment, but rather a parameterization of one?

Reply:
Yeah, `tox -l` lists the environments which are declared in `[tox] envlist = ...` at the top of the file. Unfortunately there isn't a way to say "this set of [factors](http://tox.readthedocs.io/en/latest/config.html#factors-and-factor-conditional-settings) is valid, but don't run that environment by default".